tracectl/README.md

```diff
@@ -0,0 +1,149 @@
+
+This crate exists to solve the above issues allowing:
+
+  * the automatic registration of targets across multiple crates in a centralized database, *at build time*, ensuring that:
+      * target names are *always* up-to-date in the database
+      * no target is missed
+  * programmatically chaginging of each of the target log levels at run-time
+
+## Usage model and requirements
+* This crate cannot (nor should) understand trace semantics. Trace/log semantics and their relations must be provided by the crates defining them.
+* Crates *register* their targets of interest (the targets whose loglevel is desired to be dynamically adjusted) by *declaring* them, along with their initial, default log level and the associated **tags**. Tags serve two purposes. First, they act as stable identifiers to refer to targets independently of their name; e.g. in an API. Second, a target may be associated with multiple tags. This allows controlling multiple targets simultaneously. For instance, in a packet pipeline, each network functions (NF) may emit logs to a distinct target; e.g. a  NAT NF may have a target labelled as *nat*. This may allow enabling / disabling NAT-related debug logs at runtime, while only emitting warnings or errors in production. If, in addition, the NAT (and rest of NFs) are associated with some tag *pipeline*, one may be able to enable / disable the logs (or restrict them to, say, up to INFO) in all of the NFs composing the pipeline.
+So, the takeaway is that tags represent **sets of targets** and a target can be member of an arvitrary number of sets.
+* Tags are implemented as strings since each crate should be able to define them and having a custom type (e.g. some *enum* in a centralized crate, like this one) would entail needing to update that crate every time some other crate required a new tag.
+* However, if a tag is to be shared by distinct crates, the consistency of that needs to be enforced outside of this crate.
+
+## Implementation
+This implementation has about 3 pieces:
+  * a *tracing controller*: a thread-safe database of targets, with some static initialization, that allows changing the loglevel of targets and the default one and that allows reporting which targets are available, with which tags and loglevel. This is important to be able to expose the targets in some form of API. We may expose the *tags* as stable identifiers.
+  * the automated discovery of targets at link time so that no APIs are needed to declare targets and tags.
+  * macros to ease target declaration and simplify the addition of logs
+
+# Usage
+
+## Target registration (how to declare targets in crates)
+
+### Implicit targets
+Registering implicit targets is straightforward. We just need to declare the target in the module with macro `trace_target!` to associate it with a tag and set the initial loglevel.
```

Comment:
I'm not sure I understand what the initial or default loglevel is.

Looking at the examples below, we always use `error!()` or `warn!()` etc. to produce log messages, so I expect the log level to be always set in this case. So I suppose this is not what it relates to.

Is it the default level of logs we print for this target? So if we don't configure target `T1` in the binary, and its default is `WARN`, we'll only print warnings and errors for `T1`, and `ttrace!(T1, "msg");` will be ignored?